runtime/vm/xr64/UpcallThunkGen.cpp

```diff
@@ -31,36 +31,662 @@
+	instrArray[localIdx++] = STR(2, 1, 0);
+	instrArray[localIdx++] = STR(3, 1, 8);
+	instrArray[localIdx++] = STR(4, 1, 16);
+	instrArray[localIdx++] = STR(5, 1, 24);
+	instrArray[localIdx++] = SUB(6, 6, 1);
+	instrArray[localIdx++] = ADD(0, 0, 32);
+	instrArray[localIdx++] = ADD(1, 1, 32);
+	instrArray[localIdx++] = CBNZ(6, -44); /* loop back to LDR(2, 0, 0) */
+
+	for (I_32 gIdx = 0; gIdx < roundUpSlots; gIdx++) {
+		instrArray[localIdx++] = LDR(gIdx + 2, 0, gIdx * 8);
+	}
+
+	for (I_32 gIdx = 0; gIdx < roundUpSlots; gIdx++) {
+		instrArray[localIdx++] = STR(gIdx + 2, 1, gIdx * 8);
+	}
+
+	*currIdx = localIdx;
+}
+
+/**
+ * @brief Generate the appropriate thunk/adaptor for a given J9UpcallMetaData
  *
- *     1) and 3) are mandatory, while 2) and 4) depend on the particular signature under consideration.
- *     mainly 4) implies needing 2), since this adaptor expects a return from java side before
- *     returning to the native caller.
+ * @param metaData[in/out] a pointer to the given J9UpcallMetaData
+ * @return the address for this future upcall function handle
  */
 void *
 createUpcallThunk(J9UpcallMetaData *metaData)
 {
-	// Return the thunk descriptor
-	return (void *)(&(metaData->functionPtr));
+	J9JavaVM *vm = metaData->vm;
+	const J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+	J9UpcallSigType *sigArray = metaData->nativeFuncSignature->sigArray;
+	I_32 lastSigIdx = (I_32)(metaData->nativeFuncSignature->numSigs - 1); /* The index of the return type in the signature array */
+	I_32 tempInt = 0;
+	I_32 instrCount = 0;
+	bool hiddenParameter = false;
+	bool resultDistNeeded = false;
+
+	Assert_VM_true(lastSigIdx >= 0);
+
+	/*
+	 * Instructions to call the dispatcher: 7
+	 *   set metaData to x0 (4 instructions), load target-addr, set up argListPtr, blr
+	 * Instructions to extend and restore stack frame: 5
+	 *   set up and restore SP (2 instructions), save and restore LR (2 instructions), ret
+	 */
+	instrCount = 7 + 5;
+
+	/* Testing the return type */
+	tempInt = sigArray[lastSigIdx].sizeInByte;
+	switch (sigArray[lastSigIdx].type) {
+		case J9_FFI_UPCALL_SIG_TYPE_VOID:
+			metaData->upCallCommonDispatcher = (void *)vmFuncs->native2InterpJavaUpcall0;
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_CHAR:  /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_SHORT: /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_INT32:
+			metaData->upCallCommonDispatcher = (void *)vmFuncs->native2InterpJavaUpcall1;
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_POINTER: /* Fall through */
+		case J9_FFI_UPCALL_SIG_TYPE_INT64:
+			metaData->upCallCommonDispatcher = (void *)vmFuncs->native2InterpJavaUpcallJ;
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_FLOAT:
+			metaData->upCallCommonDispatcher = (void *)vmFuncs->native2InterpJavaUpcallF;
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+			metaData->upCallCommonDispatcher = (void *)vmFuncs->native2InterpJavaUpcallD;
+			break;
+		case J9_FFI_UPCALL_SIG_TYPE_STRUCT_AGGREGATE_ALL_SP: /* all floats */
```

Comment:
I am not sure if, for ARM ABI, you need to handle specially ALL_CHAR/SHORT/INT/LONG/POINTER as ALL_SP and ALL_DP.  i.e. homogeneous float/double and homogeneous char/int etc.  such as char[], int[].  If you do, these STRUCT_AGGREGATE types may need to be added in OpenJ9.  For other platforms, these "new" types can be handled as if they were AGGREGATE_MISC or AGGREGATE_OTHER, depending on the size I guessed.

Reply:
I don't think I need to handle cases such as ALL_CHAR and ALL_SHORT like I do for ALL_SP and ALL_DP.
ALL_SP and ALL_DP with 4 elements or less is a special case in AArch64 ABI, passing the elements in FPRs.
ALL_CHAR is handled in the same way as the mixed type case.  Structs of 16 bytes or smaller are passed in two GPRs as the packed image, homogeneous or heterogeneous.  Struct `{ char c; float f; double d; }`, for example, is 16-byte long with padding between `c` and `f`, and it is passed in two GPRs.  Struct `{ int i; int j; int k; }` is also passed in two GPRs.